sdk/servicebus/Microsoft.Azure.ServiceBus/src/MessageReceivePump.cs

```diff
@@ -206,77 +228,194 @@ void CancelAutoRenewLock(object state)
             }
         }
 
-        async Task AbandonMessageIfNeededAsync(Message message)
+        async Task AbandonMessagesIfNeededAsync(IList<Message> messages)
         {
-            try
+            if (this.messageReceiver.ReceiveMode == ReceiveMode.PeekLock)
             {
-                if (this.messageReceiver.ReceiveMode == ReceiveMode.PeekLock)
+                bool batchAbandonSucceeded;
+                if (messages.Count > 1)
                 {
-                    await this.messageReceiver.AbandonAsync(message.SystemProperties.LockToken).ConfigureAwait(false);
+
+                    try
+                    {
+                        var lockTokens = messages.Select(m => m.SystemProperties.LockToken);
+
+                        await this.messageReceiver.AbandonAsync(lockTokens).ConfigureAwait(false);
+
+                        batchAbandonSucceeded = true;
+                    }
+                    catch
+                    {
+                        batchAbandonSucceeded = false;
+                    }
+                }
+                else
+                {
+                    batchAbandonSucceeded = false;
+                }
+
+                if (!batchAbandonSucceeded)
+                {
+                    foreach (var message in messages)
+                    {
+                        try
+                        {
+                            await this.messageReceiver.AbandonAsync(message.SystemProperties.LockToken).ConfigureAwait(false);
+                        }
+                        catch (Exception exception)
+                        {
+                            await this.RaiseExceptionReceived(exception, new[] { message }, ExceptionReceivedEventArgsAction.Abandon).ConfigureAwait(false);
```

Comment:
The fact we're awaiting these inside the for-loop means that it could take a long time if the user registers a long-running exception handler.  Thoughts?

Reply:
I think the consideration is that the exception handler could be doing some IO-bound work (store failure in a DB or something else). The `await` would not block everything, it would allow other IO-bound work to continue. So my assumption is that it's OK.